File: 13week/남규/2. Minimum Area Rectangle.cpp
class Solution {
public:
    unordered_map<int, unordered_map<int, int>> matrix;
    
    int minAreaRect(vector<vector<int>>& points) {
        for(auto point : points) {
            int y = point[0];
            int x = point[1];
            matrix[y][x] = 1;
        }
        
        int N = points.size();
        int ans = INT_MAX;
        
        for(int i = 0; i < N; i++){
            int y = points[i][0];
            int x = points[i][1];
            for(int j = i + 1; j < N; j++){
                int nextY = points[j][0];
                int nextX = points[j][1];
                
                if(y == nextY || x == nextX) continue;
                if(matrix[y][nextX] && matrix[nextY][x]){
                    ans = min(ans, abs(nextY - y) * abs(nextX - x));
                }
            }
        }
        return ans == INT_MAX ? 0 : ans;
    }
};

File: 13week/근우/2.minimum-area-rectangle.py
## https://leetcode.com/problems/minimum-area-rectangle/

# 한점을 기준으로 x좌표 같은놈들, y좌표가 같은놈들을 찾는다.

# 현재 점 p1
# x좌표 같은점들 모음 Set(Xp)
# y좌표 같은점들 모음 Set(Yp)

# 다시 탐색하면서 Xp와 y좌표가 있고 Yp와 x좌표가 있으면 넓이 계산을 한다.

# Timecomplexity
# *(500l+500) 2N^2


from typing import List


class Solution:
    def minAreaRect(self, points: List[List[int]]) -> int:
        min_area = NO_AREA = 2e9
        for i, p in enumerate(points):
            if p[0] == 3 and p[1] ==1:
                print()
            Xp = set()
            Yp = set()
            for j in range(i+1, len(points)):
                p2 = points[j]
                if p[0] == p2[0]:
                    Yp.add(p2[1])
                elif p[1] == p2[1]:
                    Xp.add(p2[0])
            for j in range(i+1, len(points)):
                p2 = points[j]
                if p2[0] in Xp and p2[1] in Yp:
                    min_area = min(min_area, abs((p[0] -p2[0])*(p[1]-p2[1])))
        if min_area == NO_AREA:
            return 0
        return min_area

Solution().minAreaRect([[1,1],[1,3],[3,1],[3,3],[4,1],[4,3]])